From dd1169597568785e16b7345a907e3422a6bf98b0 Mon Sep 17 00:00:00 2001
Message-ID: <dd1169597568785e16b7345a907e3422a6bf98b0.1764947809.git.lorenzo@kernel.org>
From: Lorenzo Bianconi <lorenzo@kernel.org>
Date: Fri, 5 Dec 2025 16:14:35 +0100
Subject: [PATCH] net: airoha: Improve logic to detect if the port is used as
 lan or wan

Take into account all possible port/device combinations in order to use
the port as lan or wan.

Signed-off-by: Lorenzo Bianconi <lorenzo@kernel.org>
---
 drivers/net/ethernet/airoha/airoha_eth.c | 10 +++++-----
 drivers/net/ethernet/airoha/airoha_eth.h | 23 ++++++++++++++++++++---
 drivers/net/ethernet/airoha/airoha_ppe.c |  2 +-
 3 files changed, 26 insertions(+), 9 deletions(-)

diff --git a/drivers/net/ethernet/airoha/airoha_eth.c b/drivers/net/ethernet/airoha/airoha_eth.c
index 84d9bc7e989e..951555cd9511 100644
--- a/drivers/net/ethernet/airoha/airoha_eth.c
+++ b/drivers/net/ethernet/airoha/airoha_eth.c
@@ -71,13 +71,13 @@ static void airoha_qdma_irq_disable(struct airoha_irq_bank *irq_bank,
 	airoha_qdma_set_irqmask(irq_bank, index, mask, 0);
 }
 
-static void airoha_set_macaddr(struct airoha_gdm_port *port, const u8 *addr)
+static void airoha_set_macaddr(struct airoha_gdm_dev *dev, const u8 *addr)
 {
+	struct airoha_gdm_port *port = dev->port;
 	struct airoha_eth *eth = port->qdma->eth;
 	u32 val, reg;
 
-	reg = airhoa_is_lan_gdm_port(port) ? REG_FE_LAN_MAC_H
-					   : REG_FE_WAN_MAC_H;
+	reg = airhoa_is_wan_gdm_dev(dev) ? REG_FE_WAN_MAC_H : REG_FE_LAN_MAC_H;
 	val = (addr[0] << 16) | (addr[1] << 8) | addr[2];
 	airoha_fe_wr(eth, reg, val);
 
@@ -1681,7 +1681,7 @@ static int airoha_dev_set_macaddr(struct net_device *netdev, void *p)
 	if (err)
 		return err;
 
-	airoha_set_macaddr(dev->port, netdev->dev_addr);
+	airoha_set_macaddr(dev, netdev->dev_addr);
 
 	return 0;
 }
@@ -1748,7 +1748,7 @@ static int airoha_dev_init(struct net_device *netdev)
 	u32 pse_port, fe_cpu_port;
 	u8 ppe_id;
 
-	airoha_set_macaddr(port, netdev->dev_addr);
+	airoha_set_macaddr(dev, netdev->dev_addr);
 
 	switch (port->id) {
 	case 3:
diff --git a/drivers/net/ethernet/airoha/airoha_eth.h b/drivers/net/ethernet/airoha/airoha_eth.h
index b395085fd2f8..9ed7e5ed70b0 100644
--- a/drivers/net/ethernet/airoha/airoha_eth.h
+++ b/drivers/net/ethernet/airoha/airoha_eth.h
@@ -636,13 +636,30 @@ u32 airoha_rmw(void __iomem *base, u32 offset, u32 mask, u32 val);
 #define airoha_qdma_clear(qdma, offset, val)			\
 	airoha_rmw((qdma)->regs, (offset), (val), 0)
 
-static inline bool airhoa_is_lan_gdm_port(struct airoha_gdm_port *port)
+static inline bool airhoa_is_wan_gdm_dev(struct airoha_gdm_dev *dev)
 {
+	struct airoha_gdm_port *port = dev->port;
+	struct airoha_eth *eth = port->qdma->eth;
+
 	/* GDM1 port on EN7581 SoC is connected to the lan dsa switch.
-	 * GDM{2,3,4} can be used as wan port connected to an external
+	 * GDM2 is always used as wan port is available.
+	 * GDM{3,4} can be used as lan or wan port connected to an external
 	 * phy module.
 	 */
-	return port->id == 1;
+
+	/* GDM2 is always used as wan */
+	if (port->id == 2)
+		return true;
+
+	/* GDM2 is available, this port is used as lan */
+	if (eth->ports[1])
+		return false;
+
+	/* GDM1 is always used as LAN port */
+	if (port->id == 1)
+		return false;
+
+	return !!eth->ports[0];
 }
 
 static inline bool airoha_is_7581(struct airoha_eth *eth)
diff --git a/drivers/net/ethernet/airoha/airoha_ppe.c b/drivers/net/ethernet/airoha/airoha_ppe.c
index 13ef77fb58c1..aeda3c673993 100644
--- a/drivers/net/ethernet/airoha/airoha_ppe.c
+++ b/drivers/net/ethernet/airoha/airoha_ppe.c
@@ -323,7 +323,7 @@ static int airoha_ppe_foe_entry_prepare(struct airoha_eth *eth,
 			/* For downlink traffic consume SRAM memory for hw
 			 * forwarding descriptors queue.
 			 */
-			if (airhoa_is_lan_gdm_port(port))
+			if (!airhoa_is_wan_gdm_dev(dev))
 				val |= AIROHA_FOE_IB2_FAST_PATH;
 			if (dsa_port >= 0)
 				val |= FIELD_PREP(AIROHA_FOE_IB2_NBQ,
-- 
2.52.0

